db/compaction/compaction_job.cc

```diff
@@ -290,6 +296,128 @@ void CompactionJob::Prepare() {
+  uint64_t num_planned_subcompactions =
+      std::min({static_cast<uint64_t>(ranges_size_limit),
+                max_subcompactions_limit, max_output_files_limit});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // For round-robin compaction prioity, we may have a large compaction with
+    // multiple input files in the start level. In this case, we need to employ
+    // more subcompactions (may exceed the max_subcompaction limit). The extra
+    // subcompactions will be executed using reserved threads and taken into
+    // account bg_compaction_scheduled. The number of planned subcompaction is
+    // determined through 3 steps:
+
+    // Step 1: Initialized against max_output_file_limit, the number of input
+    // files in the start level, and also the range size limit
+    num_planned_subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files_limit,
+         static_cast<uint64_t>(ranges_size_limit)});
+
+    TEST_SYNC_POINT_CALLBACK("CompactionJob::GetNumPlannedSubcompactions:Init",
+                             &num_planned_subcompactions);
+    if (max_subcompactions_limit < num_planned_subcompactions) {
+      // Step 2: Check if the initial number exceeds user-configured
+      // max_subcompaction. If yes, we need to compute the extra subcompaction
+      // based on compaction limits (the extra subcompactions should not
+      // be larger than that)
+      int extra_num_subcompactions_limit =
+          std::min((int)(num_planned_subcompactions - max_subcompactions_limit),
+                   available_bg_compactions_db_limit_);
+
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:"
+          "ExtraCompactionsAgainstDBLimits",
+          &extra_num_subcompactions_limit);
+
+      // Step 3: Try to reserve threads for extra subcompactions. The planned
+      // subcompactions may change again if we cannot obtain sufficient threads
+      extra_num_subcompaction_threads_reserved_ =
+          env_->ReserveThreads(extra_num_subcompactions_limit, thread_pri_);
+      num_planned_subcompactions =
+          static_cast<uint64_t>(extra_num_subcompaction_threads_reserved_) +
+          max_subcompactions_limit;
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:Finalize",
+          &num_planned_subcompactions);
+    }
+  }
+  return num_planned_subcompactions;
+}
+
+void CompactionJob::UpdateResourcesAndLimits(
+    uint64_t num_actual_subcompactions, uint64_t num_planned_subcompactions) {
+  // We only need to update resources and limits if the actual subcompactions
+  // is less than the planned one and we did reserve some resources earlier
+  if (extra_num_subcompaction_threads_reserved_ > 0 &&
+      num_actual_subcompactions < num_planned_subcompactions) {
+    // We cannot release threads more than what we reserved before
+    int additional_bg_compactions = env_->ReleaseThreads(
```

Comment:
maybe `int extra_num_subcompaction_threads_released = ..` so that extra_num_subcompaction_threads_reserved_ -= extra_num_subcompaction_threads_released makes more sense?

Reply:
Yes. I will change this.